src/ILLink.Shared/DataFlow/ValueSet.cs

```diff
@@ -9,49 +13,161 @@ namespace ILLink.Shared.DataFlow
 	public readonly struct ValueSet<TValue> : IEquatable<ValueSet<TValue>>, IEnumerable<TValue>
 		where TValue : notnull
 	{
-		public readonly HashSet<TValue>? Values;
+		// Since we're going to do lot of type checks for this class a lot, it is much more efficient
+		// if the class is sealed (as then the runtime can do a simple method table pointer comparison)
+		sealed class EnumerableValues : HashSet<TValue>
+		{
+			public EnumerableValues (IEnumerable<TValue> values) : base (values) { }
+
+			public override int GetHashCode ()
+			{
+				int hashCode = 0;
+				foreach (var item in this)
+					hashCode = HashUtils.Combine (hashCode, item);
+				return hashCode;
+			}
+		}
+
+		public struct Enumerator : IEnumerator<TValue>, IDisposable, IEnumerator
+		{
+			readonly object? _value;
+			int _state;  // 0 before begining, 1 at item, 2 after end
+			readonly IEnumerator<TValue>? _enumerator;
+
+			internal Enumerator (object? values)
+			{
+				_state = 0;
+				if (values is EnumerableValues valuesSet) {
+					_enumerator = valuesSet.GetEnumerator ();
+					_value = default (TValue);
+				} else {
+					_enumerator = null;
+					_value = values;
+				}
+			}
+
+			// TODO: How to get this to work - without the '!' at the end this complains that default can return null.
+			// But how does this work for HashSet<T>? That will return null from Current in reality as well... 
+			// It seems that the nullability is not propagated "inside" HashSet (since that one is implemented with TValue being nullable)
```

Comment:
Maybe the difference is that for `HashSet<T>`, the `T` isn't constrained to `notnull` like it is here.

Reply:
Yes - that is the difference. Hashset implements this without the `!` obviously. But if I use Hashset with a T which is constrained, like this:
```C#
class NonNullHashSet<T> : HashSet<T> where T : notnull { }
```

There's no warning generated by the compiler - even though the implementation is identical. (I know that the compiler can't really warn here as it doesn't see INTO the hashset).

@agocke - maybe you know about this problem some more?

I don't really mind the `!` mainly because `IEnumerator.Current` is defined as returning `default` if the enumerator is not positioned on a valid value (before start, or after end), so the functionality is correct. I'm just curious how this is supposed to work.

Reply:
It looks like the compiler does [warn](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8CwAUAMQyoA25Ew5ApgAS0zV1EDC9AxvQLz0y0A7vQDiAHjYA+egAoAlADoRtBPIDcRIgAEATPQ4BvAL7a94gCrSDRerfpaAzPXOiV83tIy0AZhAoINQiMgA) when `T` is unconstrained, in case it's instantiated with a non-nullable reference type.

And the implementation of `HashSet<T>` uses `!` I think for the same reason you did here:
https://github.com/dotnet/runtime/blob/85642f8147d9e2ba860c21baea25cbb4f12e665e/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSet.cs#L1541.

Reply:
I could swear I looked through the HashSet code and tried something similar with the T. Anyway...
Thanks a lot!